scripts/compile.ts

```diff
@@ -1,40 +1,56 @@
 import {existsSync} from 'fs';
 import * as npx from 'libnpx';
-import {join, relative} from 'path';
+import {join, resolve} from 'path';
 import {ls} from 'shelljs';
 
-const rootDir = join(__dirname, '..');
-const pkgDir = join(__dirname, '..', 'packages');
+const rootDir = resolve(join(__dirname, '..'));
+const pkgDir = join(rootDir, 'packages');
 
-const packages = ls(pkgDir);
+function project2path(proj: string, tsconfigName: string) {
+  if (tsconfigName === 'tsconfig.json') {
+    return resolve(join(pkgDir, proj));
+  } else {
+    return resolve(join(pkgDir, proj, tsconfigName));
+  }
+}
+
+async function main() {
+  const packages = ls(pkgDir);
+
+  async function compile(tsconfigName: string) {
+    console.log(`Building with ${tsconfigName}`);
 
-const tsPackages = packages.filter(path => {
-  const tsconfigPath = join(pkgDir, path, 'tsconfig.json');
-  return existsSync(tsconfigPath);
-});
+    const tsPackages = packages.filter(dir => {
+      const tsconfigPath = join(pkgDir, dir, tsconfigName);
+      return existsSync(tsconfigPath);
+    });
 
-const tsRel: string[] = [];
-const tsProjects: string[] = [];
-tsPackages.forEach(path => {
-  const basePath = join(pkgDir, path);
-  tsProjects.push(basePath);
-  tsRel.push(relative(rootDir, basePath));
-});
+    const args = ['--build', '--listEmittedFiles'];
+    if (process.argv.includes('--watch')) {
+      args.push('--watch');
+    }
 
-const args = ['-b'];
-if (process.argv.indexOf('--watch') !== -1) {
-  args.push('--watch');
+    const tsProjects: string[] = tsPackages.map(proj =>
+      project2path(proj, tsconfigName)
+    );
+
+    const res = await npx({
+      cmdOpts: args.concat(tsProjects),
+      command: 'tsc',
+      npxPkg: join(__dirname, '..', 'package.json'),
+      package: ['typescript']
+    });
+    const exit = res ? res.code : process.exitCode || 1;
+    if (exit) {
+      console.error('npx failed');
+      process.exit(1);
+    }
+  }
+
+  await compile('tsconfig.json');
+  await compile('tsconfig.next.json');
```

Comment:
@prateekbh I'm referring to this part. It's run on yarn build, but it's finishing before both functions have run

Reply:
actually it is completed here: https://travis-ci.org/prateekbh/preact-material-components/jobs/528169978#L238

Reply:
I assume it has something to do with the fact that one of the task is a promise

Reply:
Yes, the script completes, but if you Insert a `log` after `npx` it's never executed

Reply:
And I'm sure it has todo with the promise, but I didn't find a good way to prevent that and `npx` is kind of the only way, other than a child process (which is not the most portable solution), because the `typescript` compiler API doesn't work with composite builds (at least I haven't found a script/example last time looked)

Reply:
gotcha, yeah even console.log({res}) is not loggin anything. not sure if its actually waiting for npx to finish

Reply:
I know that at least the 2 npx call is never started. I'm not a node developer and haven't work much with it, so I don't know how to force node to wait. Even polls with `setTimeout` didn't work, so I started wondering if `npx` forces to process to exit somewhere and I have overlooked it? I skimmed through the code and only found it setting the Exit code, but maybe I didn't look close enough

Reply:
how about using [shelljs](https://www.npmjs.com/package/shelljs) instead for shell commands?